docs/Dialects/Debug.md

```diff
@@ -140,6 +140,45 @@ still expose the constant values under the name `Depth` and `Width` in the debug
 info.
 
 
+## Tracking Inlined Modules
+
+The `dbg.scope` op can be used to track debug information about inlined modules.
+By default, operations such as `hw.module` in conjunction with `hw.instance`
+introduce an implicit module scope. All debug operations within a module are
```

Comment:
Although I'm always a fan of making the easy case simple,  another option is to make all scopes explicit, but just add a block argument for the module's implicit (now explicit) scope.  This would mean inlining could just do a RAUW on the block arg with the explicit scope it it created for the inlined module.

Reply:
Or alternately have a `dbg.scope this` which is the scope of the current container.

Reply:
Where I'm going with this is we may have things in the future which define an implicit scope inside a module.  For example, a labeled block in verilog probably is a scope if parsing from verilog.

Reply:
Yeah absolutely. The idea of passing scopes into modules through a port/block arg is growing on me. That prevents HW ops like module and instance having to deal with debugging-specific concerns. And most passes that move stuff across the port list already know how to deal with other SSA values properly.

Right now you can model such a named Verilog block with a `dbg.scope`:
```mlir
hw.module @MyVerilogModule() {
  %0 = dbg.scope "myNamedBlock", ""
  %1 = seq.compreg ...
  dbg.variable "myRegInABlock", %1 scope %0
}
```

The absence of a `scope` operand means "current module scope". So that named Verilog block can be a `dbg.scope` op without a `scope` operand, which places it within the current module's scope. I kind of like the idea of making the scope operand non-optional, and forcing the use of a CSE-able `dbg.current_scope` or `dbg.module_scope` operation. The inliner could then just replace these ops with a corresponding `dbg.scope` op that represents the inlined module's scope.

For some of the other optimizations, like hoisting connections out of a module, it would be easier for the passes if the debug scope would always be brought into the module through the port list. Most passes that move ops across the port boundary will already rewire operands to point to the instantiation site SSA values. But we might not always want to do this, especially for public modules. That's a whole different can of worms that we'll probably want to handle separately in follow-up PRs 😬

Reply:
Although the `dbg.module_scope` could be something to add here, to avoid optional operands on the debug ops.